File: Next Greater Element - GFG/next-greater-element.js
//{ Driver Code Starts
//Initial Template for javascript
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function main() {
    let t = parseInt(readLine());
    for(let i=0;i<t;i++)
    {
        let input_line = readLine().split(' ');
        let n = parseInt(input_line[0]);
    
        input_line = readLine().split(' ');
        let arr = new Array(n);
        for(let i=0;i<n;i++){
            arr[i] = BigInt(input_line[i]);
        }
        let obj = new Solution();
        let ans = obj.nextLargerElement(arr, n);
        let s="";
        for(let i=0;i<ans.length;i++)
        {
            if(ans[i]==BigInt(-0))
                ans[i]=BigInt(0);
            s+=ans[i];
            s+=" ";
        }
        console.log(s);
    }
}

// } Driver Code Ends


//User function Template for javascript

/**
 * @param {BigInt[]} arr
 * @param {number} n
 * @returns {BigInt[]}
*/
class Solution
{
    //Function to find the next greater element for each element of the array.
    nextLargerElement(arr, n)
    {   
        let ans = new Array(n).fill(-1);
        let stack = []; // form [ [value,index] ]
        for(let i = 0; i<n; i++){
            const curr = arr[i];
            while(stack.length && curr>stack[stack.length-1][0]){
                ans[stack[stack.length-1][1]] = curr;
                stack.pop();
            }
            stack.push([curr,i]);
        }
    return ans;
    }
}

File: Next Greater Element - GFG/next-greater-element.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution
{
    public:
    //Function to find the next greater element for each element of the array.
    vector<long long> nextLargerElement(vector<long long> arr, int n){
       stack<pair<long long,int>> s; // value-index
       vector<long long> ans(n,-1);
       s.push({arr[0],0});
       for(int i = 1; i<n; i++){
           long long curr = arr[i]; // prev = s.top().first
           while(!s.empty() && curr>s.top().first){
               ans[s.top().second] = curr;
               s.pop();
           }
           s.push({curr,i});
       }
    return ans;
    }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin>>t;
    while(t--)
    {
        
        int n;
        cin>>n;
        vector<long long> arr(n);
        for(int i=0;i<n;i++)
            cin>>arr[i];
        
        Solution obj;
        vector <long long> res = obj.nextLargerElement(arr, n);
        for (long long i : res) cout << i << " ";
        cout<<endl;
    }
	return 0;
}
// } Driver Code Ends